### PR TITLE
sensors: add lsm9dsxx magnetometer driver

### PR DESCRIPTION
**Pull request by niewim19**
JIRA: PP-6

## Description

lsm9dsxx magnetometer driver code to sensorhub

## Motivation and Context

Due to hardware issues this was not tested on target, thus only draft PR is issued.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: zturn

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Comment by niewim19** [comment]
codespell check will be corrected after first review fixes

File: sensors/mag/lsm9dsxx.c [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+} lsm9dsxxMag_ctx_t;
+
+
+static int16_t translateMag(uint8_t hbyte, uint8_t lbyte)
+{
+	int16_t val = 0;
+	val |= hbyte;
+	val <<= 8;
+	val |= lbyte;
+	return val * MAG4G_CONV_MGAUSS * MGAUSS_2_UTSL * UTSL_2_MAGSHUB; /* value -> milligauss -> microtesla -> sensorhub storage scale*/
```
Comment (MaciejPurski) [inline sensors/mag/lsm9dsxx.c:93]: That would look more readable, also I think the comment is redundant, as the macros names are quite descriptive.
```suggestion
	int16_t val;

	val = (hbyte << 8) | lbyte;
 
	return val * MAG4G_CONV_MGAUSS * MGAUSS_2_UTSL * UTSL_2_MAGSHUB;
```

File: sensors/mag/lsm9dsxx.c [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+}
+
+
+static int spiWriteReg(spimsg_ctx_t *spiCtx, uint8_t regAddr, uint8_t regVal)
+{
+	unsigned char cmd[2] = { (regAddr & 0x7F), regVal }; /* write bit set to regAddr */
```
Comment (MaciejPurski) [inline sensors/mag/lsm9dsxx.c:99]: I don't understand this comment. Is it necessary?

File: sensors/mag/lsm9dsxx.c [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+#define MAG4G_CONV_MGAUSS 0.14F        /* conversion from sensor (at +-4 gauss scale) value to milligauss */
+#define MGAUSS_2_UTSL     1.221730475F /* milligauss to microtesla convertion parameter */
+#define UTSL_2_MAGSHUB    10.F         /* microtesla to sensorhub magnetic flux storage scale of 10^-7 T */
+
+
+typedef struct {
+	spimsg_ctx_t spiCtx;
+	sensor_event_t evt;
+	char stack[512] __attribute__((aligned(8)));
+} lsm9dsxxMag_ctx_t;
```
Comment (MaciejPurski) [inline sensors/mag/lsm9dsxx.c:84]: Why not `lsm9dsxx_ctx_t`?

File: sensors/mag/lsm9dsxx.c [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int lsm9dsxxMag_hwSetup(spimsg_ctx_t *spiCtx)
+{
+	if (lsm9dsxxMag_whoamiCheck(spiCtx) != 0) {
```
Comment (MaciejPurski) [inline sensors/mag/lsm9dsxx.c:122]: Why not `lsm9dsxx_whoamiCheck`?

File: sensors/mag/lsm9dsxx.c [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+		if (err >= 0) {
+			/* minus accounts for non right-handness of lsm9dsxx accelerometer frame of reference */
+			ctx->evt.mag.magX = translateMag(ibuf[1], ibuf[0]);
+			ctx->evt.mag.magY = translateMag(ibuf[3], ibuf[2]);
+			ctx->evt.mag.magZ = translateMag(ibuf[5], ibuf[4]);
+			ctx->evt.timestamp = tstamp;
+			/* TEST */
+			printf("%d %d %d\n", ctx->evt.mag.magX, ctx->evt.mag.magY, ctx->evt.mag.magZ);
+			sensors_publish(info->id, &ctx->evt);
+		}
```
Comment (MaciejPurski) [inline sensors/mag/lsm9dsxx.c:177]: What about calling `gettime()` only, when there is no error?
```suggestion
		if (err >= 0) {
			gettime(&tstamp, NULL);
			ctx->evt.timestamp = tstamp;

			/* minus accounts for non right-handness of lsm9dsxx accelerometer frame of reference */
			ctx->evt.mag.magX = translateMag(ibuf[1], ibuf[0]);
			ctx->evt.mag.magY = translateMag(ibuf[3], ibuf[2]);
			ctx->evt.mag.magZ = translateMag(ibuf[5], ibuf[4]);
			/* TEST */
			printf("%d %d %d\n", ctx->evt.mag.magX, ctx->evt.mag.magY, ctx->evt.mag.magZ);
			sensors_publish(info->id, &ctx->evt);
		}
```

**Review by MaciejPurski — CHANGES_REQUESTED** [review]
Some stylistic remarks, which require your attention.

File: sensors/mag/lsm9dsxx.c (hunk 2 above)
Comment (niewim19) [reply sensors/mag/lsm9dsxx.c:99]: Changed comment to: `/* set first transmitted bit of regAddr to 0 to indicate write operation */`. 

I thought that some random bitewise operation on register address might need explanation. Without it we might randomly read something into not existing buffer passed to xfer.

File: sensors/mag/lsm9dsxx.c (hunk 5 above)
Comment (niewim19) [reply sensors/mag/lsm9dsxx.c:177]: called `gettime` directly on event timestamp

File: sensors/mag/lsm9dsxx.c (hunk 2 above)
Comment (niewim19) [reply sensors/mag/lsm9dsxx.c]: Removed all together after consulations

**Review by MaciejPurski — DISMISSED** [review]
Looks good to me.

**Comment by niewim19** [comment]
Last changes made after testing code on hardware and finding that:
 - seeing the _SPI enable_ bit in magnetometer registers does complete opposite

+ rebase is needed

**Review by MaciejPurski — APPROVED** [review]

**Pull request merged**